### PR TITLE
ADBDEV-4567 resolve problem with AO-table resource cleaning

### PR DESCRIPTION
**Pull request by dkovalev1**
Problem: 
some GreenPlum functions, f.e. __gp_aoblkdir and __gp_aovisimap are not release
resources if used in context with limit clause.

Cause: 
It is possible for Set Returning Function (SRF) to work in ValuePerCall
mode. But according to the PostgreSQL documentation the query is not guaranteed
to be run to completion; that is, due to options such as LIMIT, the executor
might stop making calls to the set-returning function before all rows have been
fetched. This means it is not safe to perform cleanup activities in the last
call, because that might not ever happen. It's recommended to use Materialize
mode for functions that need access to external resources, such as file
descriptors.

However when running in Materialize mode SRF can consume a lot of resources
which will be not needed in case with LIMIT. F.e. __gp_aoblkdir is expected to
return 1 row for minipage. It means ~130K rows per 1TB of storage. This can lead
to overhead in ~7Mb for 1Tb of disk storage. This memory will be allocated on
segments and not used. To address this issue Greenplum introduced so called
Squelch function to States. Unfortunately Squelch is not supported for the
SRF's.

Fix: 
This patch adds support of squelching for SRFs with extension of
ValuePerCall protocol. To determine if the function called to release resources
this patch introduces macro SRF_IS_SQUELCH_CALL(). This macro shall be checked
at the very beginning of SRF.

To demonstrate work of Squelching ValuePerCall (SVPC) protocol functions
__gp_aoblkdir and __gp_aovisimap were rewritten to support SVPC.

Test for __gp_aoblkdir was added as well. Test for __gp_aovisimap was not added
due to non-trivial initialization of this test.

### CODE REVIEW

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests 7: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1130360

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA 7: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1130357

**Comment by BenderArenadata** [comment]
Failed job Regression tests with Postgres 7: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1130356

File: gpcontrib/gp_toolkit/expected/gp_toolkit_ao_funcs.out [hunk 1]
```diff
@@ -118,3 +118,17 @@ SELECT (t).* FROM (
  (0,6)   |     1 |              2 |        1 |          101 |          48 |         1
 (6 rows)
 
+-- Check that __gp_aoblkdir correctly frees resources when used with LIMIT
+-- start_ignore
+drop table if exists ao1;
```
Comment (RekGRpth) [inline gpcontrib/gp_toolkit/expected/gp_toolkit_ao_funcs.out]: The contents of the ignore block are redundant in the output.

File: src/backend/access/appendonly/appendonly_blkdir_udf.c [hunk 2]
```diff
@@ -56,6 +53,16 @@ gp_aoblkdir(PG_FUNCTION_ARGS)
...
 	Context			*context;
 
+	if (SRF_IS_SQUELCH_CALL())
+	{
+		funcctx = SRF_PERCALL_SETUP();
+		context = (Context *) funcctx->user_fctx;		
+		goto srf_done;
+	}
+
+	Oid       	aoRelOid = PG_GETARG_OID(0);
```
Comment (RekGRpth) [inline src/backend/access/appendonly/appendonly_blkdir_udf.c]: Why did you move the variable declaration from the beginning of the block?!

File: src/backend/access/appendonly/appendonly_visimap_udf.c [hunk 3]
```diff
@@ -32,6 +32,14 @@ Datum gp_aovisimap(PG_FUNCTION_ARGS);
 Datum
 gp_aovisimap(PG_FUNCTION_ARGS)
 {
+	FuncCallContext *funcctx;
+
+	if (SRF_IS_SQUELCH_CALL())
```
Comment (RekGRpth) [inline src/backend/access/appendonly/appendonly_visimap_udf.c]: I think you should move this before SRF_IS_FIRSTCALL.

File: src/backend/access/appendonly/appendonly_visimap_udf.c [hunk 4]
```diff
@@ -32,6 +32,14 @@ Datum gp_aovisimap(PG_FUNCTION_ARGS);
...
+	{
+		funcctx = SRF_PERCALL_SETUP();
```
Comment (RekGRpth) [inline src/backend/access/appendonly/appendonly_visimap_udf.c]: Maybe add it here too? `context = (Context *) funcctx->user_fctx;`?

File: src/backend/executor/nodeProjectSet.c [hunk 5]
```diff
@@ -349,3 +349,32 @@ ExecReScanProjectSet(ProjectSetState *node)
...
+	ExecSquelchNode(innerPlanState(node));
+}
+
+void
+ExecSquelchProjectSRF(ProjectSetState *node)
+{
+	int			argno;
+	ExprContext *econtext = node->ps.ps_ExprContext;
+	
+	for (argno = 0; argno < node->nelems; argno++)		{
```
Comment (RekGRpth) [inline src/backend/executor/nodeProjectSet.c]: pgindent your changes

**Comment by RekGRpth** [comment]
Remove spaces at the end of lines.

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests 7: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1133180

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests 7: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1138174

**Comment by BenderArenadata** [comment]
Failed job Behave tests 7: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1138173

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests 7: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1139574

File: src/backend/access/appendonly/appendonly_blkdir_udf.c [hunk 6]
```diff
@@ -40,28 +40,36 @@ Datum gp_aoblkdir(PG_FUNCTION_ARGS);
...
 	typedef struct Context
 	{
-		Relation 				aorel;
-		SysScanDesc 			scan;
-		MinipagePerColumnGroup	currMinipage;
-		bool					currMinipageValid;
-		int 					currMinipageEntryIdx;
-		Relation				blkdirrel;
-	} Context;
+		Relation	aorel;
```
Comment (RekGRpth) [inline src/backend/access/appendonly/appendonly_blkdir_udf.c]: You should apply pgindent to only your changes, not to other code!

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests 7: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1140299

File: src/backend/access/appendonly/appendonly_blkdir_udf.c [hunk 7]
```diff
@@ -40,7 +40,7 @@ Datum gp_aoblkdir(PG_FUNCTION_ARGS);
 Datum
 gp_aoblkdir(PG_FUNCTION_ARGS)
 {
-	Oid       	aoRelOid = PG_GETARG_OID(0);
+	Oid	aoRelOid;
```
Comment (RekGRpth) [inline src/backend/access/appendonly/appendonly_blkdir_udf.c:43]: Why not revert and this change?

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests 7: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1140701

File: src/backend/access/appendonly/appendonly_blkdir_udf.c (hunk 7 above)
Comment (dkovalev1) [reply src/backend/access/appendonly/appendonly_blkdir_udf.c:43]: It can not be initialized while squelching because function arguments are not accessible at that moment.
